tests/suites/host_test.function

```diff
@@ -432,6 +432,38 @@ static void write_outcome_result(FILE *outcome_file,
     fflush(outcome_file);
 }
 
+#if defined(__unix__) ||                                \
+    (defined(__APPLE__) && defined(__MACH__)) ||        \
+    defined(_WIN32)
+#define MBEDTLS_HAVE_CHDIR
+#endif
+
+#if defined(MBEDTLS_HAVE_CHDIR)
+/** Try chdir to the directory containing argv0.
+ *
+ * Failures are silent.
+ */
+static void try_chdir(const char *argv0) {
```

Comment:
Note: on Windows, compiled by Visual Studio,

```
printf("argv[0] = %s\n", argv0);
```

gives

```
D:\home\tom\projects\mbedtls\master\build>tests\Debug\test_suite_base64.exe
argv[0] = tests\Debug\test_suite_base64.exe
Failed to open test file: .\test_suite_base64.datax
```

or

```
D:\home\tom\projects\mbedtls\master\build>tests\Debug\test_suite_base64
argv[0] = tests\Debug\test_suite_base64
Failed to open test file: .\test_suite_base64.datax
```

i.e. `argv0` has `\`s in it

Reply:
Good point, I didn't bother with backslash support. I don't want to write complex stuff for Windows because I don't have a system to test it. I'll add a comment to document that it isn't implemented.

Reply:
Actually, I think I can do something simple that makes sense. Updated. Let me know if it doesn't work.

Reply:
I'm sure there are plenty of edge cases that don't work, such as file names containing backslashes on a Unix-like environment on Windows, or paths involving drive names. This is a developer convenience, so it doesn't have to handle all cases.

Reply:
`try_chdir()` is getting complicated enough with `#if`s that maybe we should switch to

```
diff --git a/tests/suites/host_test.function b/tests/suites/host_test.function
index 06f391fa4..9b53fd142 100644
--- a/tests/suites/host_test.function
+++ b/tests/suites/host_test.function
@@ -432,6 +432,41 @@ static void write_outcome_result(FILE *outcome_file,
     fflush(outcome_file);
 }
 
+#if defined(__unix__) ||                                \
+    (defined(__APPLE__) && defined(__MACH__))
+#define MBEDTLS_CHDIR(dir)      chdir(dir)
+#define PATH_SEPARATOR_CHAR     '/'
+#elif defined(_WIN32)
+#define MBEDTLS_CHDIR(dir)      _chdir(dir)
+#define PATH_SEPARATOR_CHAR     '\\'
+#endif
+
+#if defined(MBEDTLS_CHDIR)
+/** Try chdir to the directory containing argv0.
+ *
+ * Failures are silent.
+ */
+static void try_chdir(const char *argv0) {
+    const char *slash = strrchr(argv0, PATH_SEPARATOR_CHAR);
+    if (slash == NULL) {
+        return;
+    }
+    size_t path_size = slash - argv0 + 1;
+    char *path = mbedtls_calloc(1, path_size);
+    if (path == NULL) {
+        return;
+    }
+    memcpy(path, argv0, path_size - 1);
+    path[path_size - 1] = 0;
+    int ret = MBEDTLS_CHDIR(path);
+    if (ret != 0) {
+        mbedtls_fprintf(stderr, "%s: note: chdir(\"%s\") failed.\n",
+                        __func__, path);
+    }
+    mbedtls_free(path);
+}
+#endif /* MBEDTLS_CHDIR */
+
 /**
  * \brief       Desktop implementation of execute_tests().
  *              Parses command line and executes tests from
```

Reply:
`chdir`/`_chdir` vs slash-only/slash-or-backslash isn't necessarily the same condition since Unix-like environments on Windows can be confronted with backslashes.

Reply:
https://github.com/Mbed-TLS/mbedtls/commit/a5c8361f87921715880bca763cd9fc74f616f8f0 added just enough Windows support to get compilation errors with some mingw builds and runtime error when running unit tests compiled with MSVC and run with cmake's `ctest`. I don't understand why and I don't have the tools to reproduce the problem. So I've completely removed Windows support. If someone wants Windows support, please add it in a follow-up.